### DISCUSSION TITLE
How to end subscription obeservables properly from server?

### DISCUSSION THREAD

**Discussion by addlistener** [trpc/trpc#4060] [Q&A]
This is my server code and at some point of time I will call `emit.complete()`
```
      return observable<string>((emit) => {
        const onData = (data: string) => {
          emit.next(data);
        };
        const onError = () => {
          console.log('emit.error');
          emit.error(true);
        }
        const onComplete = () => {
          emit.complete();
        };

        // console.log('eeMap[input.emitterId]', eeMap[input.emitterId], input.emitterId);
        eeMap[input.emitterId]?.on("data", onData);
        eeMap[input.emitterId]?.on("error", onError);
        eeMap[input.emitterId]?.on("complete", onComplete);
        eeMap[input.emitterId]?.emit("registered");

        return () => {
          // console.log('off', input.emitterId);
          eeMap[input.emitterId]?.off("data", onData);
          eeMap[input.emitterId]?.off("error", onError);
          eeMap[input.emitterId]?.off("complete", onComplete);
          delete eeMap[input.emitterId];
        };
      });
    }),
```

my client code. bascially setting up some dumb listeners
```
        return tClient.testConnection.progress.subscribe({
          emitterId,
        }, {
          onData: (d) => {
            // console.log('on data', d);
            setLog((l) => l + d);
          },
          onError: (e) => {
            console.log('on error', e);
            console.error(e);
            setForceEdit(true);
          },
          onStopped: () => {
            console.log('on stopped');
          }
        });
```

But I get `TRPCClientError: Operation ended prematurely` by simply `emit.on("complete")`
<img width="744" alt="image" src="https://user-images.githubusercontent.com/4531670/227101189-95564689-8503-4dd6-8d29-abec2f5f5358.png">

So my onError listeners are triggered and the page state are updated to an error state which is undesirable. Have I done sth wrong? Isn't it normal to complete the observable on server?

I also checked the code and found `isDone` is related to the error. But I don't think it's possible to be changed on the server because `isDone` is only manipulated from the error callback.
https://github.com/trpc/trpc/blob/949f42b900291a3985c178d8043ec1ffccc9e235/packages/client/src/links/wsLink.ts#L330-L334

<img width="484" alt="image" src="https://user-images.githubusercontent.com/4531670/227139942-6cc10f85-3948-4cd7-881a-1c4d96813dcd.png">

**Comment by gunhaxxor** [comment]
Perhaps change the title? If I understand your question correct it's about closing the subscription and not the wsLink? The wsLink will be active throughout the lifetime of the TRPCProxyClient. 

Unfortunately I'm not sure why your code isn't working. Looking at it I would assume it should work. Hopefully someone else can contribute!

**Reply by addlistener** [reply]
Thanks. Updated!

**Reply by addlistener** [reply]
They would work. But the problem is that I have no way to fix the `TRPCClientError: Operation ended prematurely`. So it hard for me to distinguish if the server execution has error or not.

**Reply by addlistener** [reply]
@KATT Calling Alex for help 🥇

**Reply by PVSPHANINDRA** [reply]
@Dealerpriest Having wsLink active throughout the lifetime of the TRPCProxyClient will overload the server with the WebSocket connections.

In my application scenario, I need a web socket connection only for specific routes and I need the client to request the web socket connection only for those route requests. Since the TRPCProxyClient is making the wsLink active in the initialization, my server is getting connected irrespective of the routes.

 I raised this  #4725. It would be really helpful if you can provide me with some solution.

**Reply by KATT** [reply]
Feel free to make a pr that makes the subscription 'lazy' and connects on demand

**Comment by gunhaxxor** [comment]
I've asked about ending subscriptions from server earlier:
https://github.com/trpc/trpc/discussions/3779
Unfortunately I haven't received any suggestions. Is it perhaps so that the subscriptions currently doesn't support ending from server side? The examples in the documentation only end subscriptions from client...

**Reply by addlistener** [reply]
Haha we can do `complete` but on the client side there's always an error being thrown.

Currently I hacked it by using `emit.error` with the error message as an enum

<img width="868" alt="image" src="https://user-images.githubusercontent.com/4531670/227709463-8a3b63b8-7e21-48cc-bfe2-7022cb59fd1f.png">

**Reply by BickfordA** [reply]
@addlistener If you unsubcribe in the `onStopped` function the subscription will end (is disconnected) before the error is processed (`onError` is not called).

Like this:

```
    const subscription = await client.job.tail.subscribe(input, {
      onData(data) {
        process.stdout.write(data.toString());
        process.stderr.write('\n');
      },
      onError(error) {
        process.stderr.write(error.toString());
        process.stderr.write('\n');
      },
      onComplete() {
        process.stderr.write('Done\n');
      },
      onStopped() {
        subscription.unsubscribe();
      },
    });
```

This works following an `emit.complete()` from the server

**Reply by addlistener** [reply]
@BickfordA The problem is that, the complete event needs to be triggered from the server. In your example it's still triggered from a client unsub if I understand you correctly.

I can always make a complex workaround but it's just less desirable.

**Reply by BickfordA** [reply]
@addlistener The `emit.complete()` is triggered from the server in my example.  

Try adding the unsubcribe to your onStopped example. I think this should prevent the `onError` from being called as the connection is closed before it is processed. 
```ts
  const sub =  tClient.testConnection.progress.subscribe({
      emitterId,
    }, {
      onData: (d) => {
        // console.log('on data', d);
        setLog((l) => l + d);
      },
      onError: (e) => {
        console.log('on error', e);
        console.error(e);
        setForceEdit(true);
      },
      onStopped: () => {
        sub.unsubscribe();
        console.log('on stopped');
      }
    });
  return sub;
```

**Comment by lamualfa** [comment]
`emit.complete` also doesn't work for me using `ipcLink` from `electron-trpc`. So I guess the problem comes from the `@trpc/server`.

<img width="284" alt="image" src="https://github.com/trpc/trpc/assets/39755201/b5f97cb8-c375-4cfb-bac7-74ed20de9d74">

**Edit**

After troubleshooting for a while, I found that the bug comes from the closing mechanism on the client.

https://github.com/jsonnull/electron-trpc/pull/174
